contrib/firebaseio/src/main/java/com/google/cloud/dataflow/contrib/firebase/utils/FirebaseCheckpoint.java

```diff
@@ -65,15 +64,15 @@ public void finalizeCheckpoint() throws IOException {
   }
 
   public Instant minTimestamp(){
-    if (cur.compareTo(unread.peek()) < 0){
+    if (unread.isEmpty() || cur.compareTo(unread.peek()) < 0){
```

Comment:
not sure this is threadsafe -- could be interrupted at the `||`. It looks like `peek` returns `null` if the collection is empty, so I'd suggest:

```java
@Nullable Record<T> next = unread.peek();
if (next == null || cur.compareTo(next) < 0) {
    return cur.timestamp;
} else {
    return next.timestamp;
}
```

(I may have already asked this, but is `cur.compareTo(next)` the right comparison or should you be doing comparison on their timestamps? If the former is sufficient, can you add a comment as to why?

e.g., 

```java
// note that cur.compareTo(next) implicitly compares timestamps as well
```